### PR TITLE
chore: expand presets in `synthesizer-ui`

### PR DESCRIPTION
**Pull request by wodeni**
# Description

Related issue/PR: #1133 

This PR follows up #1133 and adds in the rest of the geometry examples to the `synthesizer-ui` presets. The PR also fixes a few small bugs in the user interface.

# Implementation strategy and design decisions

* Add 13 additional presets to `synthesizer-ui` from `geometry-domain`.
* Change multi-select download behavior for `Grid`
* Check type safety for `SwapIn` and `Swap` mutations

# Examples with steps to reproduce them

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new ESLint warnings
- [x] I have reviewed any generated changes to the `diagrams/` folder

### CODE REVIEW

File: packages/core/src/analysis/SubstanceAnalysis.ts [hunk 1]
```diff
@@ -253,23 +267,24 @@ export const identicalTypeDecls = (
...
-  const idSet = ids.map((i) => i.value);
+  const idStrs = ids.map((i) => i.value);
   let options: im.Map<string, Identifier<A>[]> = im.Map();
+  log.debug(`Matching ${ids.map((i) => i.value)}`);
 
-  idSet.forEach((i) => {
-    const obj = env.vars.get(i);
-    const typeStr = obj && obj.name.value;
+  idStrs.forEach((i) => {
+    const typeStr = env.vars.get(i)?.name.value;
```
Comment (samestep) [inline packages/core/src/analysis/SubstanceAnalysis.ts:275]: What's the motivation behind this change?

File: packages/core/src/analysis/SubstanceAnalysis.ts [hunk 2]
```diff
@@ -318,6 +333,8 @@ export const cascadingDelete = <T>(
         return findArg(s, id);
       } else if (s.tag === "Decl") {
         return s.name === id;
+      } else if (s.tag === "LabelDecl") {
+        return s.variable.value === id.value;
```
Comment (samestep) [inline packages/core/src/analysis/SubstanceAnalysis.ts:337]: Would it make sense to use `switch` for this instead of `if`/`else`? That would allow us to automatically check exhaustivity.

File: packages/core/src/synthesis/Synthesizer.ts [hunk 3]
```diff
@@ -399,9 +463,10 @@ export class Synthesizer {
       )
     );
     // add autolabel statement
-    this.updateProg(autoLabel(this.currentProg));
+    // TODO: find out what to label
+    // this.updateProg(autoLabel(this.currentProg));
     return {
-      prog: this.currentProg,
+      prog: sortStmts(dedupStmts(this.currentProg)), // sort and deduplicate statements to make sure edits don't introduce compiler errors
```
Comment (samestep) [inline packages/core/src/synthesis/Synthesizer.ts:469]: What does this mean exactly? Is the idea that object declarations always get sorted before their usages in constructors/functions/predicates?

**Review by samestep — APPROVED** [review]
Looks solid! Left a few comments. There was some stuff in `packages/core/src/synthesis/Synthesizer.ts` for which I felt like I was missing context; could you walk me through it in person tomorrow, perhaps? Also, I was wondering, is there a reason that the code in `packages/core/src/{analysis,synthesis}/` doesn't live in `packages/synthesizer/` instead?

File: packages/synthesizer-ui/package.json [hunk 4]
```diff
@@ -53,6 +53,9 @@
     "react-dom": "^18.2.0"
   },
   "devDependencies": {
+    "@types/file-saver": "^2.0.5",
+    "@types/jszip": "^3.4.1",
+    "typescript": "^4.5.4",
```
[the remark this answers is not in the document]
Comment (wodeni) [reply packages/synthesizer-ui/package.json:58]: yeah these are used for the "export all" button

File: packages/synthesizer-ui/src/main.tsx [hunk 5]
```diff
@@ -5,7 +5,7 @@ import "./index.css";
 
 ReactDOM.render(
   <React.StrictMode>
-    <App />
+    <App />,
   </React.StrictMode>,
```
[the remark this answers is not in the document]
Comment (wodeni) [reply packages/synthesizer-ui/src/main.tsx:9]: umm interesting that it went through `typecheck`. But no this was not intentional

File: packages/examples/src/geometry-domain/textbook_problems/c11p25.substance [hunk 6]
```diff
@@ -1,11 +1,15 @@
...
+Segment CO := MkSegment(C, O)
+EqualLengthMarker(AO, CO)
 Segment AB := Chord(c, A, B)
 Segment DC := Chord(c, D, C)
-Segment DB := Chord(c, D, B)
-Segment AC := Chord(c, A, C)
 Segment CB := Chord(c, B, C)
 Segment DA := Chord(c, D, A)
+Segment DB := Diameter(c, D, B)
+Segment AC := Diameter(c, A, C)
```
[the remark this answers is not in the document]
Comment (wodeni) [reply packages/examples/src/geometry-domain/textbook_problems/c11p25.substance:12]: The original sub program was written wrong basically. The diagram is constructed with `AC` and `DB` being diameters of the circle, not chords. 

Untitled Diagram (2)

File: packages/core/src/analysis/SubstanceAnalysis.ts (hunk 2 above)
Comment (wodeni) [reply packages/core/src/analysis/SubstanceAnalysis.ts:337]: Good catch! Helped me handle a few more cases too.

File: packages/core/src/analysis/SubstanceAnalysis.ts (hunk 1 above)
Comment (wodeni) [reply packages/core/src/analysis/SubstanceAnalysis.ts:275]: Mainly to clarify that this is not really a set, but a list to match with existing type constructors.

**Review by samestep — APPROVED** [review]
I'm cool to merge this!
